### PR TITLE
Fix deprecated CHARSET utf8 in install and upgrade

### PR DESCRIPTION
**Pull request by leemyongpakvn**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | install script prior <s>5.0.1</s> 6.0.0 version CREATE TABLE with deprecated CHARSET=utf8. It need to be updated to utf8mb4 in install and upgrade process of the next version.
| Type?         | improvement
| BC breaks?    | no
| Deprecations? | yes
| Fixed ticket? | Fixes PrestaShop/Prestashop#28857
| How to test?  | Check the above ticket.

### CODE REVIEW

**Review by kpodemski — APPROVED** [review]

**Comment by kpodemski** [comment]
Hey @leemyongpakvn 👋 

PHP-CS-Fixer is probably not happy with an indentation you have in your upgrade file

**Comment by leemyongpakvn** [comment]
@kpodemski PHP-CS-Fixer v2 complains about 1 redundant space. I have no problem with PHP-CS-Fixer v3 on my localhost :yum:

**Review by kpodemski — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @leemyongpakvn ,

I have an issue, I have installed version 5.0.0 of productcomments module.
But I don't have the upgrade to v5.0.3 available, only v5.0.2:

Screenshot 2022-09-28 at 09 47 54

Could you check ?
Thanks!

**Comment by kpodemski** [comment]
@florine2623 I think it's because you need to bump the version of the module in `productcomments.php` 🤔

**Comment by okom3pom** [comment]
@florine2623 

You must create version 5.0.3. Say me if you need help

**Comment by okom3pom** [comment]
Ha ! @kpodemski again too fast

**Comment by florine2623** [comment]
@kpodemski @okom3pom I though it would be automatic !

Can one of you guy show me how to do it ?

I changed it in productcomments.php line 48 `$this->version = '5.0.3';`. Not sure if that's the right way to do it 😅

**Comment by okom3pom** [comment]
Here you have two things to test: 

The update and the installation.

You have to check that the tables will be in utf8mb4

Yes, you have to change the version at this line

**Comment by florine2623** [comment]
Here's what I did:

- git clone git@github.com:PrestaShop/productcomments.git
- click Install on BO
- git checkout tags/v5.0.0
- Install PR: git prc origin 144 dev
- Modified productcomments.php line 48 `$this->version = '5.0.3';`
- Checked BO, upgrade available only to 5.0.2. Not 5.0.3.

Am I missing something ?

**Comment by leemyongpakvn** [comment]
@kpodemski  & @okom3pom According to [modules/creation/enabling-auto-update](https://devdocs.prestashop-project.org/1.7/modules/creation/enabling-auto-update/), the upgrade filename must be **upgrade**-5.0.3.php. Upgraded successfully on PS 1.7.8.5 but failed on PS 8.0.0 with filename `upgrade`-5.0.3.php.
Amazingly other previous upgrade files of this modules are **install**-`version`.php :thinking:

**Comment by leemyongpakvn** [comment]
@florine2623 Please change filename from **install**-5.0.3.php to **upgrade**-5.0.3.php in module's **upgrade** folder also, then try to upgrade again.

**Comment by leemyongpakvn** [comment]
I have other more strange problem on PS 8.0.0. After clone module with `git clone https://github.com/PrestaShop/productcomments.git`, then install from BO, I still see **5.0.1** in BO although in productcomments.php line 48 $this->version = '**5.0.2**' 
Change productcomments.php line 48 to $this->version = '**5.0.3**'; will lead to this strange screen
PS8 0_productcomments_5 0 2
It looks like the old version string **5.0.1** (in my case) is kept somewhere out of productcomments.php

**Comment by kpodemski** [comment]
I'll ping @atomiix as it seems that it might be related to the Distribution API Client

**Comment by kpodemski** [comment]
QA, the workaround is to disable `ps_distributionapi` before upgrading the module.

Ensure that module tables are all with a new encoding after upgrading to 5.0.3. Remember about changing 5.0.2 to 5.0.3 in `productcomments.php`

**Comment by leemyongpakvn** [comment]
@florine2623 Which MySQL version do you use ? On my side, MySQL 8.0.31 identified old `utf8_general_ci` as `utf8mb3_general_ci`

**Comment by sallemiines** [comment]
Hello @leemyongpakvn 

I tried to check the pr , But I don't have the upgrade to v5.0.3 available, only v5.0.2 ! 
I use the MySQL version  5.7.37
phpMyAdmin Version : 5.1.3
and I always have this result  

image

ping @PrestaShop/qa-team  Could someone try to reproduce this PR please ?

**Comment by leemyongpakvn** [comment]
@sallemiines You and **florine2623** have the same problem with the Distribution API Client. You can see discussion about how to reproduce in Fixed PR 28857. Maybe we need a few months more to see the issue easier and continue working on this PR :laughing:

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @leemyongpakvn 

I have checked your PR. I found these issues : 

Steps to test : 
```
git clone git@github.com:PrestaShop/productcomments.git
git checkout tags/v5.0.0
git pull --rebase origin dev
git prc origin 144 dev
composer install
Changed `productcomments.php` file on line 48 $this->version = '5.0.3';

```
**Check on PS1788 :** 

**-Issue 1 :** The available upgrade is v5.0.2 and not v5.0.3 : NOK :x: 
image

-After upgrade the module, the version is updated to **5.0.3** in `Modules` page : OK :heavy_check_mark:  
image

In `productcomments.php` file is well 5.0.3 and not updated
image

**-Issue 3 :** The `upgrade` button is always displayed after upgrading the module: NOK :x: 
image

-In Database table, **the utf8mb4_general_ci** is displayed in COLLATION column : OK :heavy_check_mark: 
image

https://user-images.githubusercontent.com/92912932/207881862-3e9d4493-e1c5-443e-8ff3-aba82c43fe4b.mp4



**Check on PS8.0.0 :** 

-Same of issue 1 of PS1788 : NOK :x: 
image

-The point 2 of PS1788 is NOK with PS8.0.0. The v5.0.2 is always still displayed in `Modules` page after upgrading the module : NOK :x: 
image

In `productcomments.php` file is changed to 5.0.2 after upgrading the module
image


-The issue 3 of PS1788 is not present with PS8.0.0. The `Configure` button is well displayed after upgrading the module : OK :heavy_check_mark: 
image

-In Database table, the utf8_general_ci is displayed in COLLATION column : NOK :x: 
image

https://user-images.githubusercontent.com/92912932/207882098-5c1c5fd2-2964-4be7-a7de-94ebca836f75.mp4

Could you please check it?

Thank You!

**Comment by kpodemski** [comment]
@PrestaShop/qa-team 

can a dev test it? I'm positive that this change is legitimate, problems are because updates are not coming from the git version of the module, but from the addons marketplace and distribution API client

**Comment by Robin-Fischer-PS** [comment]
@kpodemski Yes, you can test it, it will be easier indeed. Thanks ! :)

**Review by davidglezz — APPROVED** [review]

File: upgrade/install-5.0.3.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_criterion` CHARACTER SET = utf8mb4 COLLATE utf8mb4_general_ci;')
+        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_criterion_product` CHARACTER SET = utf8mb4 COLLATE utf8mb4_general_ci;')
+        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_criterion_lang` CHARACTER SET = utf8mb4 COLLATE utf8mb4_general_ci;')
+        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_criterion_category` CHARACTER SET = utf8mb4 COLLATE utf8mb4_general_ci;')
+        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_grade` CHARACTER SET = utf8mb4 COLLATE utf8mb4_general_ci;')
+        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_usefulness` CHARACTER SET = utf8mb4 COLLATE utf8mb4_general_ci;')
+        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_report` CHARACTER SET = utf8mb4 COLLATE utf8mb4_general_ci;');
```
Comment (kpodemski) [inline upgrade/install-5.0.3.php:39]:
```suggestion
    return Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;')
        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_criterion` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;')
        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_criterion_product` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;')
        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_criterion_lang` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;')
        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_criterion_category` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;')
        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_grade` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;')
        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_usefulness` CHARACTER SET = utf8mb4 COLLATE utf8mb4_general_ci;')
        && Db::getInstance()->execute('ALTER TABLE `' . _DB_PREFIX_ . 'product_comment_report` CHARACTER SET = utf8mb4 COLLATE utf8mb4_general_ci;');
```

Comment (kpodemski) [reply upgrade/install-5.0.3.php:39]: @leemyongpakvn could you take a look? this works for me, and it upgrades all the fields in the table, your code didn't work on MySQL 5.7

**Comment by leemyongpakvn** [comment]
@kpodemski Although `product_comment_usefulness` and `product_comment_report` tables don't have char/text fields, I think they still need `CONVERT TO` like others. 
Another issue is `COLLATION` keywords at the end of every `CREATE TABLE` commands in [install-dev](https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/data/db_structure.sql#L10). I don't know why they are there but I guess we need do the same for this module **install.php**.
Be free to add suggestions if you think they are needed :)

**Comment by davidglezz** [comment]
I think there is a corner case:

User have old version installed and then uninstalls it keeping the tables.
User install new version.

I think that migration scrip is not executed so, that user will still having `utf8` instead of `utf8mb4`

I thin that on install we should check if tables exists.

Wow, this is a bigger problem, if the user decides to keep the old version tables, the columns might be different in a new version.

**Comment by leemyongpakvn** [comment]
@davidglezz I see parameter `$keep = true` in [productcomments.php's function uninstall](https://github.com/PrestaShop/productcomments/blob/v6.0.0/productcomments.php#L107-L110), 
but in Back-Office it looks like keep the module's **files**
> Optional: delete module folder after uninstall.

not keep the module's **tables**.
If a user knows how to delete the module's files **manually** while keep its tables, then reinstall module, he/she must be a dev or a QA :fearful:

**Comment by leemyongpakvn** [comment]
Fortunately, this `install-6.0.1.php` still runs when testing in a dev style (remove module folder, git clone, apple this PR changes, change module version to `6.0.1`  in productcomments.php and config.xml, then upgrade via BO). A bit funny when I checked Module Upgrade in BO
productcomments_600_to_601

but Upgrade button did its task correctly: module tables COLATTION changed to `utf8mb4`, and module version become `6.0.1`.

**Comment by leemyongpakvn** [comment]
Thanks @kpodemski and @nicosomb

**Comment by leemyongpakvn** [comment]
It is a bit surprised that `install-6.0.3.php` still works with `function upgrade_module_6_0_1` inside ;)

**Pull request merged**